### PR TITLE
C++: Disambiguate SideEffectFunction QLDoc.

### PR DESCRIPTION
**Pull request by geoffw0**
Disambiguate the `SideEffectFunction` QLDoc as we discussed in the meeting on Thursday (we may want to add models for input/output/access to 'hidden' memory at some point, but for now, we are content to make the meaning of what we have unambiguous).

I've also corrected what I believe to be an error in `PureStrFunction`.

### CODE REVIEW

**Comment by rdmarsh2** [comment]
LGTM, but the change to `PureFunction` affects the IR tests.

**Comment by geoffw0** [comment]
Ah, good point.  I've rebased these changes (because a lot has changed on master recently) and added the test changes.  It looks like these are limited to `CallReadSideEffect` disappearing from certain (pure) function calls, which sounds about right.

File: cpp/ql/src/semmle/code/cpp/models/interfaces/SideEffect.qll [hunk 1]
```diff
@@ -18,14 +18,18 @@ abstract class SideEffectFunction extends Function {
   /**
    * Holds if the function never reads from memory that was defined before entry to the function.
    * This memory could be from global variables, or from other memory that was reachable from a
-   * pointer that was passed into the function.
+   * pointer that was passed into the function. Input side-effects, and reads from memory that
+   * cannot be visible to the caller (for example a buffer inside an I/O library) are not modelled
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/models/interfaces/SideEffect.qll]:
```suggestion
   * cannot be visible to the caller (for example a buffer inside an I/O library) are not modeled
```

Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/SideEffect.qll]: TIL this is a British vs. American English difference

File: cpp/ql/src/semmle/code/cpp/models/interfaces/SideEffect.qll [hunk 2]
```diff
@@ -18,14 +18,18 @@ abstract class SideEffectFunction extends Function {
...
    */
   abstract predicate hasOnlySpecificReadSideEffects();
 
   /**
    * Holds if the function never writes to memory that remains allocated after the function
    * returns. This memory could be from global variables, or from other memory that was reachable
-   * from a pointer that was passed into the function.
+   * from a pointer that was passed into the function. Output side-effects, and writes to memory
+   * that cannot be visible to the caller (for example a buffer inside an I/O library) are not
+   * modelled here.
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/models/interfaces/SideEffect.qll]:
```suggestion
   * modeled here.
```

**Pull request merged**